Parse log levels from cw entries

sgronblo opened the pull request:
Adds a new option parse_log_levels to the Terraform module. When this value is set to "true" the log entries will be parsed to see if the original log message string was prefixed with a marker "[LOG_LEVEL]", the LOG_LEVEL will be parsed out and used as the log level when sending entries to Papertrail.

sam-gronblom-rj committed:
Add simple log level parsing from prefix marker

- Checks whether a log entry starts with "[${LOG_LEVEL}]" and then uses
that for setting the level when sending logs to Papertrail
- The log level parsing is controlled using the "parse_log_levels"
Terraform module input variable, which should be "true" or "false"
- Added a small test for the log level parsing along with testing
dependencies
- If log level parsing is turned off or if no log level can be parsed,
the log level will default to "info"

sam-gronblom-rj committed:
Add .editorconfig file to set indentation to 2

sam-gronblom-rj committed:
Remove test folder from compiled files

- Add more informative output to script

sam-gronblom-rj committed:
Update compiled lambda-code

sam-gronblom-rj committed:
Handle tab separator in log messages

sam-gronblom-rj committed:
Update Lambda source code files

sam-gronblom-rj committed:
Remove warning for non-TSV messages

- Lambda itself will log rows that don't match this pattern as part of
normal operation: eg. "START RequestId:
6adea095-e601-11e7-bbcd-c346de6c314d Version: $LATEST"
- It would spam up the log for this Lambda to print a warning for each
of these

sam-gronblom-rj committed:
Update Lambda code

sam-gronblom-rj committed:
Merge branch 'master' into parse-log-levels-from-cw-entries

uhef commented on cloudwatch-papertrail.ts:
```diff
@@ -46,9 +46,29 @@ function getEnvVarOrFail(varName: string): string {
+    return undefined
+  }
+  const message = messageColumns[2]
+  const match = logLevelRegex.exec(message)
```
We can combine both the split to date, request, message and log level parsing to single regex:
`/^.+\t.+\t\[(\w+)\]/` or something like this...

uhef replied:
Perhaps make it stricter than above...

sgronblo replied:
Fixed

uhef commented on prepare-module.sh:
```diff
@@ -10,14 +10,18 @@ set -u
 cp package.json lambda-files
 cd lambda-files || exit "Directory 'lambda-files' does not exist"
+echo "* Installing Javascript dependencies for Lambda"
 npm install --no-package-lock --production
```
isn't installing without package lock dangerous? can this lead into situation where we have different dependency versions when testing locally and when deploying to production? perhaps we can avoid this situation by copying the package lock over as well.

sgronblo replied:
Yes, that is true. Fixed.

uhef commented on prepare-module.sh:
```diff
@@ -10,14 +10,18 @@ set -u
+echo "* Installing Javascript dependencies for Lambda"
 npm install --no-package-lock --production
+echo "* Removing test folder and package.json from Lambda source files"
+rm -rf test
```
This removes `test` folder from dist folder, I assume. How did the `test` folder end up into the dist folder in the first place?

sgronblo replied:
The test folder ends up there because we are currently using the same tsconfig.json file both for the actual Lambda code and its test code. So when you run `tsc` you end up with the compiled test files in the outDir as well. I think we could avoid this by having a separate tsconfig.json for the tests but didn't feel like spending time setting it up when I could just simply remove the compiled tests file in the `prepare-module.sh` script.

uhef commented on tf_module/lambda-files/cloudwatch-papertrail.js:
```diff
@@ -24,9 +24,29 @@ function getEnvVarOrFail(varName) {
     }
     return value;
 }
+// Should match for example: "[error] The database has exploded"
```
This is a build artifact file, right? Should it be in the version control system at all?

sgronblo replied:
Yes, this is a constraint imposed on us by the following things:

1. We are relying on Github (instead of something like S3) for hosting the Terraform module files
2. We want to impose as few dependency requirements on the end users of the module as possible. (Ie. users of the module should not need to have Docker or typescript or npm, etc installed to build the compiled artifacts)

It seems like the least complicated choice at this time. If we had a reliable S3 bucket to store the Lambda source files we could get rid of having to store the compiled artifacts in source control.

uhef replied:
Right. This is fine for now. Would another option be to use the terraform module registry: https://registry.terraform.io/

sam-gronblom-rj committed:
Use package-lock.json when installing production dependencies

sam-gronblom-rj committed:
Use combined RegExp to get level without explicit pre-splitting

sam-gronblom-rj committed:
Change default log format to be based on winston's simple format

- It makes more sense to start with the default and change to a custom
format or add additional formats when we know that it is necessary

Commit:
Fix comments
